src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -28,249 +29,847 @@
+             && (new_partition == ShenandoahFreeSetPartitionId::Collector)) ||
+            ((orig_partition == ShenandoahFreeSetPartitionId::Collector)
+             && (new_partition == ShenandoahFreeSetPartitionId::Mutator)))) ||
+          ((available == _region_size_bytes) &&
+           ((orig_partition == ShenandoahFreeSetPartitionId::Mutator)
+            && (new_partition == ShenandoahFreeSetPartitionId::Collector))), "Unexpected movement between partitions");
+
+  size_t used = _region_size_bytes - available;
+
+  _membership[int(orig_partition)].clear_bit(idx);
+  _membership[int(new_partition)].set_bit(idx);
+
+  _capacity[int(orig_partition)] -= _region_size_bytes;
+  _used[int(orig_partition)] -= used;
+  shrink_interval_if_boundary_modified(orig_partition, idx);
+
+  _capacity[int(new_partition)] += _region_size_bytes;;
+  _used[int(new_partition)] += used;
+  expand_interval_if_boundary_modified(new_partition, idx, available);
+
+  _region_counts[int(orig_partition)]--;
+  _region_counts[int(new_partition)]++;
+}
+
+const char* ShenandoahRegionPartitions::partition_membership_name(idx_t idx) const {
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
+  ShenandoahFreeSetPartitionId result = ShenandoahFreeSetPartitionId::NotFree;
+  for (uint partition_id = 0; partition_id < NumPartitions; partition_id++) {
+    if (_membership[partition_id].is_set(idx)) {
+      assert(result == ShenandoahFreeSetPartitionId::NotFree, "Region should reside in only one partition");
+      result = (ShenandoahFreeSetPartitionId) partition_id;
+    }
+  }
+  return partition_name(result);
+}
+
+
+#ifdef ASSERT
+inline ShenandoahFreeSetPartitionId ShenandoahRegionPartitions::membership(idx_t idx) const {
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
+  ShenandoahFreeSetPartitionId result = ShenandoahFreeSetPartitionId::NotFree;
+  for (uint partition_id = 0; partition_id < NumPartitions; partition_id++) {
+    if (_membership[partition_id].is_set(idx)) {
+      assert(result == ShenandoahFreeSetPartitionId::NotFree, "Region should reside in only one partition");
+      result = (ShenandoahFreeSetPartitionId) partition_id;
+    }
+  }
+  return result;
+}
+
+inline bool ShenandoahRegionPartitions::partition_id_matches(idx_t idx, ShenandoahFreeSetPartitionId test_partition) const {
+  assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT, idx, _max);
+  assert (test_partition < ShenandoahFreeSetPartitionId::NotFree, "must be a valid partition");
+
+  return membership(idx) == test_partition;
+}
+#endif
+
+inline bool ShenandoahRegionPartitions::is_empty(ShenandoahFreeSetPartitionId which_partition) const {
+  assert (int(which_partition) < NumPartitions, "selected free partition must be valid");
+  return (leftmost(which_partition) > rightmost(which_partition));
+}
+
+inline idx_t ShenandoahRegionPartitions::find_index_of_next_available_region(
```

Comment:
If I read this correctly, calling code assumes that the result is >= start_index, right? Otherwise loops might not terminate. Maybe assert that this holds?

Reply:
Good idea.  Done.